tools/clang/lib/Sema/SemaHLSL.cpp

```diff
@@ -2103,16 +2112,16 @@ static bool CombineBasicTypes(ArBasicKind LeftKind,
                               ArBasicKind RightKind,
                               _Out_ ArBasicKind* pOutKind)
 {
-  if ((LeftKind < 0 || LeftKind >= AR_BASIC_COUNT) ||
-    (RightKind < 0 || RightKind >= AR_BASIC_COUNT)) {
-    return false;
-  }
-
   if (LeftKind == RightKind) {
     *pOutKind = LeftKind;
     return true;
   }
 
+  if ((LeftKind < 0 || LeftKind >= AR_BASIC_COUNT) ||
+    (RightKind < 0 || RightKind >= AR_BASIC_COUNT)) {
+    return false;
+  }
+
```

Comment:
I would think validation takes higher priority than the equal type behavior above. Why was this if block moved under?

Reply:
The reason is because the objects are not "BASIC". So even though they matched, they were failing due to this. I could maintain some of the validation though.